Feat: fastify support with nitro (#386)
vercel/workflow: docs/content/docs/getting-started/fastify.mdx

---
title: Fastify
---

This guide will walk through setting up your first workflow in a Fastify app. Along the way, you'll learn more about the concepts that are fundamental to using the development kit in your own projects.

---

<Steps>

<Step>
## Create Your Fastify Project

Start by creating a new Fastify project.

```bash
mkdir my-workflow-app
```

Enter the newly made directory:

```bash
cd my-workflow-app
```

Initialize the project:

```bash
npm init --y
```

### Install `workflow`, `fastify` and `nitro`

```package-install
npm i workflow fastify nitro rollup
```

<Callout>

> Fastify doesn’t ship with a build system. Nitro provides one so you can compile workflows and run them in development and production. Learn more about Nitro [here](https://v3.nitro.build).

Reviewer: ```suggestion
By default, Fastify doesn't include a build system. Nitro adds one which enables compiling workflows, runs, and deploys for development and production. Learn more about [Nitro](https://v3.nitro.build).
```

</Callout>

> If using TypeScript, add types:

Reviewer: ```suggestion
If using TypeScript, you need to install the `@types/node` and `typescript` packages
```

```bash
npm i -D @types/node typescript
```

### Configure Nitro

> Create `nitro.config.ts` with the `workflow/nitro` module. This enables the `"use workflow"` and `"use step"` directives.

Reviewer: ```suggestion
Create a new file `nitro.config.ts` for your Nitro configuration with module `workflow/nitro`. This enables usage of the `"use workflow"` and `"use step"` directives
```

```typescript title="nitro.config.ts" lineNumbers
import { defineNitroConfig } from "nitro/config";

export default defineNitroConfig({
	modules: ["workflow/nitro"],
	vercel: { entryFormat: "node" },

> 	preset: "vercel",

Reviewer: fyi i think nitro auto-detects presets depending on the environment variables set, so i don't think this is needed

	routes: {

> 		"/**": { handler: "./src/index.ts" },

Author: for some reason, this causes an issue locally for me both with express and fastify. removing "format":"node" and adding preset: "node-server" | "vercel" | "node" works

<img width="1150" height="654" alt="image" src="https://github.com/user-attachments/assets/bf2b579a-55aa-4bc0-8ab1-599892370622" />

    </AccordionTrigger>
    <AccordionContent className="[&_p]:my-2">
To enable helpful hints in your IDE, set up the workflow plugin in `tsconfig.json`:

```json title="tsconfig.json" lineNumbers
{
  "compilerOptions": {
    // ... rest of your TypeScript config
    "plugins": [
      {
        "name": "workflow" // [!code highlight]
      }
    ]
  }
}
```

    </AccordionContent>
  </AccordionItem>
</Accordion>

### Update `package.json`

Add Nitro scripts:

```json title="package.json" lineNumbers
{
  // ...
  "scripts": {
    "dev": "nitro dev",
    "build": "nitro build"
  },
  // ...
}
```

</Step>

<Step>

## Create Your First Workflow

Create a new file for our first workflow:

```typescript title="workflows/user-signup.ts" lineNumbers
import { sleep } from "workflow";

export async function handleUserSignup(email: string) {
  "use workflow"; // [!code highlight]

  const user = await createUser(email);
  await sendWelcomeEmail(user);

  await sleep("5s"); // Pause for 5s - doesn't consume any resources
  await sendOnboardingEmail(user);

  return { userId: user.id, status: "onboarded" };
}
```

We'll fill in those functions next:

> 

Author: lords work mate 🙏 thank you!

    throw new Error("Retryable!");
  }
}

async function sendOnboardingEmail(user: { id: string; email: string }) {
  "use step"; // [!code highlight]
  if (!user.email.includes("@")) {
    // FatalError skips retries
    throw new FatalError("Invalid Email");
  }
  console.log(`Sending onboarding email to user: ${user.id}`);
}
```

</Step>

<Step>

## Create Your Route Handler

Wire up a Fastify route to start the workflow at `src/index.ts`:

```typescript title="src/index.ts"
import Fastify from "fastify";
import { fromNodeHandler } from "nitro/h3";
import { start } from "workflow/api";
import { handleUserSignup } from "../workflows/user-signup.js";

const app = Fastify({ logger: true });
app.post("/api/signup", async (req, reply) => {
  const { email } = req.body as { email: string };
  await start(handleUserSignup, [email]);
  return reply.send({ message: "User signup workflow started" });
});

// Wait for Fastify to be ready before handling requests
await app.ready();

// Expose Fastify app as a Nitro handler
export default fromNodeHandler((req, res) => {
	return new Promise((resolve) => {
		res.on("finish", resolve);
		app.server.emit("request", req, res);
	});
});
```

> 

Reviewer: ```suggestion
This route handler creates a `POST` request endpoint at `/api/signup` that will trigger your workflow.
```

</Step>

<Step>

## Run in development

Start the dev server from your Fastify app root:

```bash
npm run dev
```

Trigger the workflow:

```bash
curl -X POST --json '{"email":"hello@example.com"}' http://localhost:3000/api/signup
```

Check the Fastify logs to see the workflow orchestration and steps. You can also use the [Workflow DevKit CLI or Web UI](/docs/observability) to inspect runs:

```bash
npx workflow inspect runs # add '--web' for an interactive Web based UI
```

<img src="/o11y-ui.png" alt="Workflow DevKit Web UI" />

</Step>

> </Steps>

Reviewer: something like this (but make sure this is actually valid and no adapters etc are needed to deploy fastify on Vercel):

```suggestion
</Steps>

---

## Deploying to production

Workflow DevKit apps currently work best when deployed to [Vercel](https://vercel.com/home) and needs no special configuration.

Check the [Deploying](/docs/deploying) section to learn how your workflows can be deployed elsewhere.

## Next Steps

- Learn more about the [Foundations](/docs/foundations).
- Check [Errors](/docs/errors) if you encounter issues.
- Explore the [API Reference](/docs/api-reference).

```

  "use step"; // [!code highlight]
  if (!user.email.includes("@")) {
    // FatalError skips retries
    throw new FatalError("Invalid Email");
  }
  console.log(`Sending onboarding email to user: ${user.id}`);
}
```
Taking a look at this code:

- Business logic lives inside **steps**. When a step is invoked inside a **workflow**, it gets enqueued to run on a separate request while the workflow is suspended, just like `sleep`.
- If a step throws an error, like in `sendWelcomeEmail`, the step will automatically be retried until it succeeds (or hits the step's max retry count).
- Steps can throw a `FatalError` if an error is intentional and should not be retried.

<Callout>
  We'll dive deeper into workflows, steps, and other ways to suspend or handle
  events in [Foundations](/docs/foundations).
</Callout>
</Step>

<Step>

## Create Your Route Handler

To invoke your new workflow, we'll create both the Fastify app and a new API route handler at `src/index.ts` with the following code:

```typescript title="src/index.ts"
import Fastify from "fastify";
import { start } from "workflow/api";
import { handleUserSignup } from "../workflows/user-signup.js";

> const app = Fastify({ logger: true });

Author: logger: true is unnecessary here tbh and makes the console log noisy

Author: nvm. fastify installation docs shows to keep it, so I'll go with that
